classes/class-wc-connect-cart-validation.php

```diff
@@ -0,0 +1,145 @@
+<?php
+/**
+ * Validation logic.
+ *
+ * @package WooCommerce\Shipping
+ */
+
+/**
+ * Validates the shipping info in the cart.
+ */
+class WC_Connect_Cart_Validation {
+
+	/**
+	 * Needed to keep track of the current package being proccessed.
+	 *
+	 * @var array
+	 */
+	private $current_package;
+
+	/**
+	 * Register filters when required instead of using constructor.
+	 */
+	public function register_filters() {
+		add_filter( 'woocommerce_cart_no_shipping_available_html', [ $this, 'error_no_shipping_available_html' ] );
+		add_filter( 'woocommerce_shipping_package_name', [ $this, 'set_current_package' ], 10, 3 );
+	}
+
+	/**
+	 * We use this filter to store the current package array because it's not passed to the woocommerce_cart_no_shipping_available_html filter.
+	 *
+	 * @param string $package_name Package Name.
+	 * @param int    $i Index.
+	 * @param array  $package Current package we need to store.
+	 *
+	 * @return mixed
+	 */
+	public function set_current_package( $package_name, $i, $package ) {
+		$this->current_package = $package;
+		return $package_name;
+	}
+
+	/**
+	 * More friendly error message when WCS has an error.
+	 *
+	 * @param string $error_html Generic error message from WC.
+	 * @return string
+	 */
+	public function error_no_shipping_available_html( $error_html ) {
+		try {
+			$this->validate_package_destination( $this->current_package );
+			return $error_html;
+		} catch ( WC_Connect_Cart_Exception $exception ) {
+			return $exception->getMessage();
+		}
+	}
+
+	/**
+	 * Called by above filter to get error message to show and by shipping method to detect errors.
+	 *
+	 * @param array $package Current Package.
+	 * @return bool
+	 * @throws WC_Connect_Cart_Exception Thrown when validation error occurs.
```

Comment:
Yeah, it doesn't need to use an exception. I like to avoid using mixed types. I was leaning towards refactoring to pass on error array bey reference to the second parameter.

Reply:
I feel that by returning `null` on success and the error's string on error, then we could leverage the null coalescing operator and have things like:
```
public function error_no_shipping_available_html( $error_html ) {
  return $this->validate_package_destination( $this->current_package ) ?? $error_html;
}
...
public function is_valid_package_destination( $package ) {
  $this->get_cart_validator()->validate_package_destination( $package ) ?? false;
}
```

Which IMO could shorten things and we'd also get rid of the extra exception. The exception seems to be used just to store/wrap the error message, rather than to stop the flow because something exceptional occurred.
What do you think?